design/decisions/0005-entity-registration-api.md

```diff
@@ -0,0 +1,266 @@
+  This results in unnecessary duplication of data and makes it error-prone with multiple sources of truth.
+  A central component like `tedge-agent` would be a better fit to store and maintain the entity store.
+* The current entity store also maintains the external id of devices, which is a cloud-specific aspect.
+  When the entity store is moved to the `tedge-agent`, this aspect must be detached and left with the mappers,
+  without the mappers having to maintain a duplicate entity store.
+* When the `tedge-agent` handles the registration of entities, it can't consult with the mappers to make sure that
+  the entities that it accepted locally are valid in the cloud as well.
+  If the local registration succeeds, but the cloud registration fails, the user must get that feedback as well.
+* A mapper might get connected much later, after a lot of local registrations are already processed.
+  Even in such cases, the mapper should be able to fetch all the entities that were registered till then,
+  and replicate that in the cloud.
+  Even in that case, any failed registrations must be conveyed to the user asynchronously.
+
+## Solution Proposal
+
+### Proposal 1: REST APIs for registration
+
+* The `tedge-agent` maintains the entity store and provides REST APIs for registration, and querying.
+  The mappers and other components can query entity data from it when needed.
+* On every successful registration, `tedge-agent` publishes those registration messages to their respective topics,
+  so that other components likes mappers subscribed to those topics are notified of those new entities.
+* The `tedge-agent` can persist the entity store on disk, as it is done today, for faster recovery on restart.
+  It also helps in differentiating already seen entities from newly published ones.
+* The mappers no longer maintains the entire entity store but only maintains a simple mapping
+  between topic ids and their cloud external ids, so that mapped messages can be published to the same.
+* The mappers can rebuild their external id map on every restart, or even persist the same on disk for faster recovery.
+* When the registration of a local entity fails in the cloud for reasons like conflicting external ids,
+  the corresponding error message from the cloud is forwarded to `te/errors` topic so that the user is aware.
+
+#### Create new entity
+
+**Endpoint**
+
+```
+POST /v1/entity
+```
+
+**Payload**
+
+```json
+{
+    "@topic-id": "device/child01//",
+    "@type": "child-device",
+    "@id": "child01",
+    "name": "child01",
+    "extra-fragment": {
+        "extra-key": "extra-value"
+    }
+}
+```
+
+**Responses**
+
+* 201: Created
+  ```json
+  {
+      "@topic-id": "device/child01//"
+  }
+  ```
+* 409: Conflict
+  ```json
+  {
+      "error": "Entity with topic-id: 'device/child01//' already exists."
+  }
+  ```
+
+#### Update existing entity
+
+**Endpoint**
+
+```
+PUT /v1/entity/{topic-id}
```

Comment:
Since the `{topic-id}` includes forward slashes, this *may* cause some implementation problems (though it depends on the path rules on your HTTP handler from whatever library will be used). So it might be worth checking if that is possible first.

The alternative would be to force users to url encode the `{topic-id}`, e.g. `%2F` but that would be annoying to use.

Reply:
Yeah, I was expecting URL encoded values only. Isn't that widely used in query parameters and hence an accepted convention?

Reply:
URL encoding would be a _bit_ annoying to use, although I think this is a relatively simple case where `.replace("/", "%2F")` would generally suffice. I'm fairly sure allowing people to use non-escaped entity IDs would be technically possible, but that would prevent us from doing anything of the form `/v1/entity/{topic-id}/something-else` (or at least I would have reservations about doing that, even if it might be technically feasible).

Reply:
The pain of having to enforce url encoding of the `{topic-id}` could be reduced via a tedge cli command, `tedge register device device/child01//`...then it can do the encoding

Reply:
Doing any url encoding (in the path) is super painful if you have to do it in bash...

```
rawurlencode() {
  local string="${1}"
  local strlen=${#string}
  local encoded=""
  local pos c o

  for (( pos=0 ; pos<strlen ; pos++ )); do
     c=${string:$pos:1}
     case "$c" in
        [-_.~a-zA-Z0-9] ) o="${c}" ;;
        * )               printf -v o '%%%02x' "'$c"
     esac
     encoded+="${o}"
  done
  echo "${encoded}"    # You can either set a return variable (FASTER) 
  REPLY="${encoded}"   #+or echo the result (EASIER)... or both... :p
}
```

Reply:
Or a slightly better one (taken from the same link as above):

```sh
ENCODED_TOPIC_ID=$(curl -s -o /dev/null -w %{url_effective} --get --data-urlencode "device/main//" localhost:1883 | cut -d'?' -f2-)

```

Reply:
If URL encoding is not a preferred option, then we can reconsider using entity id directly in the path.

> I'm fairly sure allowing people to use non-escaped entity IDs would be technically possible

It should be possible as we're doing the same in the file transfer service where we allow users to POST and GET file names of arbitrary length

> but that would prevent us from doing anything of the form /v1/entity/{topic-id}/something-else

Although I can't think of any API requirements that would add further path segments after the entity id, at least for the current requirements, I don't want to rule out future evolutions of the API that would require it. But, even in that case we have a way around it, as the entity topic IDs are fixed length (at most 4 path segments). So, even if we introduce such an API in future, we can take advantage of this fixed length. The only niggle there would be to use empty path segments for things like `device/main///` which may not be a commonly accepted convention.

I wasn't planning to enforce 4 segment entity IDs for the proposed APIs, and let users just provide just `device/main` instead of `device/main///` in the path, and the server logic to append any missing trailing `/`s to compute the entity topic id, to keep it simple. We can enforce the 4-segment length only when a future API enhancement absolutely requires it.

Reply:
> Although I can't think of any API requirements that would add further path segments after the entity id

That statement didn't age well :wink:. Already came across this requirement for bulk deletion of child entities of a given entity `topic-id`.

Reply:
>> Although I can't think of any API requirements that would add further path segments after the entity id

> That statement didn't age well :wink:. Already came across this requirement for bulk deletion of child entities of a given entity `topic-id`.

I see things the other way around. Since its more convenient to have the entity topic identifier used as a suffix (to re-conciliate MQTT and HTTP naming rues), we can take this as a starting point for URL design, e.g. `/v1/entity/{topic-id}` This is also aligned with the REST principles where an URL identifies an entity and not some action.